Spigot-Server-Patches/0424-Rate-limit-packets-incoming-from-players.patch

```diff
@@ -0,0 +1,104 @@
+index a7673dd49..7c4b45522 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -462,4 +462,11 @@ public class PaperConfig {
+         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
+         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
+     }
++
++    public static int packetRateLimit = 250; // per second
++    public static String packetRateLimitKickMessage = "Sent too many packets";
++    private static void packetRateLimit() {
++        packetRateLimit = Math.max(-1, getInt("settings.spam-limiter.packets-per-second", packetRateLimit));
++        packetRateLimitKickMessage = getString("messages.kick.packet-spam", packetRateLimitKickMessage);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
+index 30f646e42..de3fb824a 100644
+--- a/src/main/java/net/minecraft/server/NetworkManager.java
++++ b/src/main/java/net/minecraft/server/NetworkManager.java
+@@ -153,6 +153,11 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     private static <T extends PacketListener> void a(Packet<T> packet, PacketListener packetlistener) {
++        // Paper start - Ratelimit packets
++        if (packetlistener instanceof PlayerConnection) {
++            ((PlayerConnection)packetlistener).rateLimitPacket(packet);
++        }
++        // Paper end
+         packet.a((T) packetlistener); // CraftBukkit - decompile error
+     }
+ 
+diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
+index dc8c20efb..5ef493188 100644
+--- a/src/main/java/net/minecraft/server/PlayerConnection.java
++++ b/src/main/java/net/minecraft/server/PlayerConnection.java
+@@ -141,6 +141,51 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+         return (this.player == null) ? null : (CraftPlayer) this.player.getBukkitEntity();
+     }
+     // CraftBukkit end
++    // Paper start - ratelimit packets
++    private static final int PACKET_LIMIT_TIME_INTERVAL = 5; // seconds
++    private final Object rateLimitPacketsLock = new Object();
++    private long rateLimitPacketsTime = System.nanoTime();
++    private int rateLimitPacketCount; // packets read since rateLimitPacketsTime
++
++    void rateLimitPacket(final Packet<?> packet) {
++        /*
++         * We cannot rely on tick() being called. If a server lags we could end up kicking everyone!
++         * So we must rely on our own timer.
++         */
++        final int limit = com.destroystokyo.paper.PaperConfig.packetRateLimit * PACKET_LIMIT_TIME_INTERVAL;
++        if (limit < 0) {
++            return;
++        }
++
++        // Note: SeqLock would perform better here if needed
++        synchronized (this.rateLimitPacketsLock) {
```

Comment:
we should be able to do this with Atomic Integers instead of synchronizing? 
other rate limit stuff is already done as atomic operations such as chat spam.

Reply:
That's because those methods have the main thread decrement them. If we rely on the main thread, then it's possible for the main thread to stall and for us to kick everyone. I don't think that's remotely desirable.

I could look into doing this lockless but the solution isn't pretty and less maintainable (at least if I want to avoid race conditions)